src/dmd/expressionsem.d

```diff
@@ -5879,6 +5876,15 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 exp.msg = new StringExp(Loc.initial, buf.extractSlice());
             }
         }
+
+        // https://issues.dlang.org/show_bug.cgi?id=20353
+        // Do not optimize before generating the message as the condition
+        // may be const-folded into a boolean.
+        //
+        // BUG: see if we can do compile time elimination of the Assert
+        exp.e1 = exp.e1.optimize(WANTvalue);
+        exp.e1 = exp.e1.toBoolean(sc);
```

Comment:
This change should be done only if the checkaction switch is enabled. If it's not, the current behavior should be kept. This may also get rid of the tester failures.

Reply:
This code above is only executed for `checkaction=context` so the code path is not changed without it.
```
/* Optimize */
if (false) { /* Rewrite message */ }
```
to 

```
if (false) { /* Rewrite message */ }
/* Optimize */
```